### PR TITLE
[OpenCV] Add MSMF & MSMF-DXVA as a Option

### PR DESCRIPTION
**Pull request by yekmen**
Specify library name and version:  opencv/4.5.5

Add as option to OpenCV:
- Build with OpenCV World 
- Build with or without msmf (Microsoft Media Foundation) support

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/opencv/4.x/conanfile.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ class OpenCVConan(ConanFile):
     options = {
         "shared": [True, False],
         "fPIC": [True, False],
+        "build_world":[True, False],
```
Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:28]: It's not a very good name like all recipe options `build_<something>`, I think`world` is a better name and consistent with other option `dnn` to build opencv_dnn.

File: recipes/opencv/4.x/conanfile.py [hunk 2]
```diff
@@ -86,6 +90,8 @@ class OpenCVConan(ConanFile):
         "with_imgcodec_pfm": False,
         "with_imgcodec_pxm": False,
         "with_imgcodec_sunraster": False,
+        "with_msmf": False,
+        "with_msmf_dxva": False,
```
Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:94]: these options should be removed if os is not Windows.

File: recipes/opencv/4.x/conanfile.py [hunk 3]
```diff
@@ -732,7 +743,11 @@ def add_components(components):
                 # TODO: we should also define COMPONENTS names of each target for find_package() but not possible yet in CMakeDeps
                 #       see https://github.com/conan-io/conan/issues/10258
                 self.cpp_info.components[conan_component].set_property("cmake_target_name", cmake_target)
-                self.cpp_info.components[conan_component].libs = [lib_name]
+                if self.options.build_world and conan_component == "opencv_world":
+                    self.cpp_info.components[conan_component].libs = [lib_name]
+                elif not self.options.build_world:
+                    self.cpp_info.components[conan_component].libs = [lib_name]
+
```
Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:750]: this branching seems to be useless, why this modification?

Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:750]: Are you sure that all other targets are defined in official OpenCV config file when world is built?

File: recipes/opencv/4.x/conanfile.py (hunk 1 above)
Comment (yekmen) [reply recipes/opencv/4.x/conanfile.py:28]: Agree, I will changed it.

File: recipes/opencv/4.x/conanfile.py (hunk 2 above)
Comment (yekmen) [reply recipes/opencv/4.x/conanfile.py:94]: Agree, I have not possibility to test on other operation system.

File: recipes/opencv/4.x/conanfile.py (hunk 3 above)
Comment (yekmen) [reply recipes/opencv/4.x/conanfile.py:750]: That why I add this condition, opencv_world will combine all generated dll with only one dll which called opencv_world.dll

When world is disable, all dll will be created, tested on my side and working well.
I will recheck with the new version of the conanfile.py

Comment (SpaceIm) [reply recipes/opencv/4.x/conanfile.py:750]: I know, but:
- all other components part of world should depend on opencv_world. They shouldn't have requires, system libs or frameworks themselves, everything should be aggregated in requires/system_libs/frameworks of opencv_world component
- 2 extra modules are never part of world: img_hash & sfm (and its static dependencies correspondence, multiview and numeric)
- ippiw is not part of world

Comment (SpaceIm) [reply recipes/opencv/4.x/conanfile.py:750]: I've updated https://github.com/conan-io/conan-center-index/pull/15193 to add `world` option.

**Comment by yekmen** [comment]
I remove opencv world which is added on your pull request now. 
I only keep msmf option

File: recipes/opencv/4.x/conanfile.py [hunk 4]
```diff
@@ -126,6 +130,8 @@ def config_options(self):
         if self.settings.os != "Linux":
             del self.options.with_gtk
             del self.options.with_v4l
+            del self.options.with_msmf
+            del self.options.with_msmf_dxva
```
Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:134]:
```suggestion
        if self.settings.os != "Windows":
            del self.options.with_msmf
            del self.options.with_msmf_dxva
```

File: recipes/opencv/4.x/conanfile.py [hunk 5]
```diff
@@ -430,8 +436,8 @@ def generate(self):
         tc.variables["WITH_EIGEN"] = self.options.with_eigen
         tc.variables["HAVE_QUIRC"] = self.options.with_quirc  # force usage of quirc requirement
         tc.variables["WITH_DSHOW"] = is_msvc(self)
-        tc.variables["WITH_MSMF"] = is_msvc(self)
-        tc.variables["WITH_MSMF_DXVA"] = is_msvc(self)
+        tc.variables["WITH_MSMF"]  = self.options.with_msmf
+        tc.variables["WITH_MSMF_DXVA"] = self.options.with_msmf_dxva
```
Comment (SpaceIm) [inline recipes/opencv/4.x/conanfile.py:440]:
```suggestion
        tc.variables["WITH_MSMF"]  = self.options.get_safe("with_msmf", False")
        tc.variables["WITH_MSMF_DXVA"] = self.options.get_safe("with_msmf_dxva", False)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

**Comment by uilianries** [comment]
@yekmen Thank you for your contribution, please, check the account `Yavuz EKMEN`, it should sign the CLA as well, because both emails pushed commits to this PRs.

**Comment by yekmen** [comment]
> @yekmen Thank you for your contribution, please, check the account `Yavuz EKMEN`, it should sign the CLA as well, because both emails pushed commits to this PRs.
Done,

**Comment by uilianries** [comment]
> Conan v1 pipeline ❌
> Sorry, the build is only launched for Access Request users. You can request access https://github.com/conan-io/conan-center-index/issues/4.

@yekmen please, ask for access request on the issue #4, so the CI will be able to run your PR.

**Comment by yekmen** [comment]
> > Conan v1 pipeline ❌
> > Sorry, the build is only launched for Access Request users. You can request access #4.
> 
> @yekmen please, ask for access request on the issue #4, so the CI will be able to run your PR.

Done, Waiting about approvement

**Comment by SpaceIm** [comment]
Could also refer to OpenCV in the title please? There are ~1500 recipes in conan-center, so this title is not very informative.

**Review by SpaceIm — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Comment by yekmen** [comment]
Waiting for the last approvement

**Review by SSE4 — APPROVED** [review]

**Comment by SpaceIm** [comment]
CI is red, you can close/open to trigger it again.

**Comment by yekmen** [comment]
According to the CI error, don't know why my PR failed. 

Error : 
Graph error
    Package 'ffmpeg/4.4' not resolved: Unable to find 'ffmpeg/4.4' in remotes
ERROR: Package 'ffmpeg/4.4' not resolved: Unable to find 'ffmpeg/4.4' in remotes

But ffmpeg is available on remote
https://conan.io/center/ffmpeg?version=4.4

Any help is welcome

**Pull request merged**